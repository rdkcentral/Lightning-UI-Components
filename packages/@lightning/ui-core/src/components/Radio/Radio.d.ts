import lng from '@lightningjs/core';
import Base from '../Base';
import type { Color, StylePartial } from '../../types/lui';

export type RadioStyle = {
  alpha: number;
  backgroundColor: Color;
  backgroundColorChecked: Color;
  knobColor: Color;
  knobHeight: number;
  knobWidth: number;
  radius: lng.Tools.CornerRadius;
  strokeColor: Color;
  strokeWidth: number;
};

export default class Radio extends Base {
  checked?: boolean;
  get style(): RadioStyle;
  set style(v: StylePartial<RadioStyle>);

  // tags 
  get _Knob(): lng.Component; 
  get _Body(): lng.Component;
  get _Stroke(): lng.Component;
}