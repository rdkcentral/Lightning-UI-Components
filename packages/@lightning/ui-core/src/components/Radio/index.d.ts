import Radio, { RadioStyles } from './Radio';
import RadioSmall from './RadioSmall';

export { Radio as default, RadioSmall, RadioStyles };
