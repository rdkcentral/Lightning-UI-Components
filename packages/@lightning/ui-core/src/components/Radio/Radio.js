import Base from '../Base/index.js';
import lng from '@lightningjs/core';
import * as styles from './Radio.styles.js';
import { withExtensions } from '../../mixins/index.js';

class Radio extends Base {
  static get __componentName() {
    return 'Radio';
  }

  static get __themeStyle() {
    return styles;
  }

  static _template() {
    const center = { mount: 0.5, x: w => w / 2, y: h => h / 2 };
    return {
      Body: {
        rtt: true, // ensures the background color doesn't bleed into the knob when disabled
        ...center,
        Knob: {
          ...center,
          alpha: 0
        }
      },
      Stroke: center
    };
  }

  static get tags() {
    return ['Knob', 'Body', 'Stroke'];
  }
  static get properties() {
    return ['checked'];
  }

  _update() {
    this._updateBody();
    this._updateStroke();
    this._updateKnob();
    if (this._checkedChanged) {
      this.fireAncestors('$announce', this.announce);
      this._checkedChanged = false;
    }
    this._updateOpacity();
  }

  _updateBody() {
    const bodyColor = this.checked
      ? this.style.backgroundColorChecked
      : this.style.backgroundColor;

    // if the inner body should be square, a rounded corner radius can still be applied to the stroke
    const radius =
      this.style.radius >= this.w / 2
        ? (this.w - this.style.strokeWidth) / 2
        : 0;

    this._Body.patch({
      texture: lng.Tools.getRoundRect(
        this.w - this.style.strokeWidth,
        this.h - this.style.strokeWidth,
        radius,
        null,
        null,
        true,
        bodyColor
      )
    });
  }

  _updateStroke() {
    this._Stroke.patch({
      texture: lng.Tools.getRoundRect(
        this.w,
        this.h,
        this.style.radius,
        this.style.strokeWidth,
        this.style.strokeColor,
        false
      )
    });
  }

  _updateKnob() {
    this._Knob.patch({
      texture: lng.Tools.getRoundRect(
        this.style.knobWidth,
        this.style.knobHeight,
        this.style.knobWidth / 2,
        null,
        null,
        true,
        this.style.knobColor
      )
    });

    this.applySmooth(this._Knob, {
      alpha: this.checked ? 1 : 0
    });
  }

  _updateOpacity() {
    this.applySmooth(this, { alpha: this.style.alpha });
  }

  _setChecked(checked) {
    this._checkedChanged = checked !== this._checked;
    return checked;
  }

  toggle() {
    if (!this._isDisabledMode) {
      this.checked = !this.checked;
    }
    return this;
  }

  _handleEnter() {
    if (typeof this.onEnter === 'function') {
      return this.onEnter(this);
    } else {
      this.toggle();
    }
    return false;
  }

  set announce(announce) {
    super.announce = announce;
  }

  get announce() {
    return this._announce || (this.checked ? 'Checked' : 'Unchecked');
  }
}

export default withExtensions(Radio);
