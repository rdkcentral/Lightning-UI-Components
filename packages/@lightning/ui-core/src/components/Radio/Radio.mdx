import { Canvas, Story } from '@storybook/addon-docs';
import { Theme } from '../../../storybook/index.js';
import Radio from './index.js';

# Radio

A radio has 2 states: "on" or "off". When used in a group, it should let a user select only one of the choices at a time. This can be managed with the withSelections mixin

## Source

https://github.comcast.com/Lightning/lightning-ui/packages/@lightning/ui-core/src/components/Radio/Radio.js

## Usage

Simple component that just needs a type.

```js
import { Radio } from '@lightning/ui';

class Example extends lng.Component {
  static _template() {
    return {
      Radio: {
        type: Radio
      }
    };
  }
}
```

<Canvas>
  <Story id="elements-radio--basic" />
</Canvas>

## API

### Properties

| name    | type    | required | default | description  |
| ------- | ------- | -------- | ------- | ------------ |
| checked | boolean | false    | false   | toggle state |

### Style Properties

| name                   | type   | description                            |
| ---------------------- | ------ | -------------------------------------- |
| backgroundColor        | string | background color of the Radio          |
| backgroundColorChecked | string | color of Radio background when checked |
| knobColor              | string | color of the Radio knob                |
| knobHeight             | number | height of knob                         |
| knobWidth              | number | width of knob                          |
| radius                 | number | border radius                          |
| strokeColor            | string | color of the Radio border              |
| strokeWidth            | number | border width                           |

### Methods

#### toggle(): void

Toggles `checked` state and updates UI accordingly

### Styles

Styled components can be adjusted by composing using [withStyles](?path=/docs/mixins-withstyles--basic). Below is the default `styles` object

<Theme theme={Radio.styles} />
