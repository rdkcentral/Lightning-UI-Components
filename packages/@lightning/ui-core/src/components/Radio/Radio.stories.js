import lng from '@lightningjs/core';
import { default as RadioComponent } from './index.js';
import { createModeControl } from '../../../storybook/index.js';
import { CATEGORIES } from 'lightning-ui-docs';

export default {
  title: `${CATEGORIES[512]}/Radio`
};

export const Radio = () =>
  class Radio extends lng.Component {
    static _template() {
      return {
        Radio: {
          type: RadioComponent
        }
      };
    }
  };
Radio.args = { checked: false };
Radio.argTypes = {
  ...createModeControl(['disabled']),
  checked: {
    control: 'boolean',
    description: 'Toggles checked between on and off'
  }
};
