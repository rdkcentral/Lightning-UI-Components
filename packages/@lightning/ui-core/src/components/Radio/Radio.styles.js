export const base = theme => {
  const size = theme.spacer.xxl;
  return {
    alpha: 1,
    w: size,
    h: size,
    knobHeight: size / 2,
    knobWidth: size / 2,
    radius: size / 2,
    strokeWidth: theme.stroke.sm
  };
};

export const palette = theme => ({
  neutral: {
    backgroundColor: theme.color.fillNeutralDisabled,
    backgroundColorChecked: theme.color.fillNeutral,
    knobColor: theme.color.fillInverse,
    strokeColor: theme.color.strokeInverse
  },
  inverse: {
    backgroundColor: theme.color.fillInverseDisabled,
    backgroundColorChecked: theme.color.fillInverse,
    knobColor: theme.color.fillNeutral,
    strokeColor: theme.color.strokeNeutral
  },
  brand: {
    backgroundColor: theme.color.fillInverseDisabled,
    backgroundColorChecked: theme.color.fillBrand,
    knobColor: theme.color.fillNeutral,
    strokeColor: theme.color.strokeNeutral
  }
});

export const mode = theme => ({
  disabled: { alpha: theme.alpha.inactive }
});
