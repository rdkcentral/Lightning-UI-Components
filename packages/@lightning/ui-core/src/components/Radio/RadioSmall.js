import Radio from './Radio.js';
import { withExtensions } from '../../mixins/index.js';
import * as styles from './RadioSmall.styles.js';

class RadioSmall extends Radio {
  static get __componentName() {
    return 'RadioSmall';
  }

  static get __themeStyle() {
    return styles;
  }
}

export default withExtensions(RadioSmall);
