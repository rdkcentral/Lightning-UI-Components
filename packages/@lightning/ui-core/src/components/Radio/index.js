export { default as default } from './Radio.js';
export { default as RadioSmall } from './RadioSmall.js';
