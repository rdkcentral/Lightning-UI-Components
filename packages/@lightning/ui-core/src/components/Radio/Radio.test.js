import { makeCreateComponent } from '@lightning/ui-test-utils';
import Radio from './index.js';
import { jest } from '@jest/globals';

const createRadio = makeCreateComponent(Radio);

describe('Radio', () => {
  let radio, testRenderer;

  beforeEach(() => {
    [radio, testRenderer] = createRadio();
  });
  afterEach(() => {
    radio = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('sets the announce string to the checked status', () => {
    radio.checked = true;
    testRenderer.forceAllUpdates();
    expect(radio.announce).toBe('Checked');
    radio.checked = false;
    testRenderer.forceAllUpdates();
    expect(radio.announce).toBe('Unchecked');
  });

  it('overrides the announce string', () => {
    const overrideString = 'Custom announce string';
    radio.announce = overrideString;
    testRenderer.forceAllUpdates();
    expect(radio.announce).toBe(overrideString);
  });

  it('reannounces on checked status change', () => {
    radio.fireAncestors = jest.fn();
    radio.checked = true;
    testRenderer.forceAllUpdates();
    expect(radio.fireAncestors).toHaveBeenCalledWith('$announce', 'Checked');
  });

  it('toggles checked state on enter', () => {
    [radio, testRenderer] = createRadio({ checked: false });
    expect(radio.checked).toEqual(false);
    testRenderer.keyPress('Enter');
    expect(radio.checked).toEqual(true);
    testRenderer.keyPress('Enter');
    expect(radio.checked).toEqual(false);
  });

  it('should not toggle the radio when in disabled mode', () => {
    // NOTE: The checked status can be set before the mode is changed
    [radio, testRenderer] = createRadio({ checked: false, mode: 'disabled' });
    expect(radio.checked).toBe(false);
    expect(radio._Knob.alpha).toBe(0);
    testRenderer.forceAllUpdates();
    testRenderer.keyPress('Enter');
    expect(radio.checked).toBe(false);
    expect(radio._Knob.alpha).toBe(0);
  });
});
