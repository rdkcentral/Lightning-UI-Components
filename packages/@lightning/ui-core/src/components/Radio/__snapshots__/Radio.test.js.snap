// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders 1`] = `
{
  "Component": {
    "Knob": {
      "active": false,
      "alpha": 0,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": false,
      "flex": false,
      "flexItem": false,
      "h": 16,
      "isComponent": undefined,
      "mount": 0.5,
      "mountX": 0.5,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Knob",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Knob",
      ],
      "texture": {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 16,
      "x": 16,
      "y": 16,
      "zIndex": 0,
    },
  },
}
`;
